lxd/ubuntupro/client.go

```diff
@@ -0,0 +1,261 @@
+	GuestAttachSettingOff GuestAttachSetting = "off"
+
+	// GuestAttachSettingAvailable indicates that guest attachment is available.
+	// - When the host has this setting, devlxd requests to `GET /1.0/ubuntu-pro` should return the setting from the guest
+	//   (`ubuntu_pro.guest_attach) and `POST /1.0/ubuntu-pro/token` should retrieve a guest token via the Ubuntu Pro client.
+	// - When the guest has this setting, the pro client inside the guest will not try to retrieve a guest token at startup
+	//   (though attachment with a guest token can still be performed with `pro auto-attach`.
+	GuestAttachSettingAvailable GuestAttachSetting = "available"
+
+	// GuestAttachSettingOn indicates that guest attachment is on.
+	// - When the host has this setting, devlxd requests to `GET /1.0/ubuntu-pro` should return the setting from the guest
+	//   (`ubuntu_pro.guest_attach) and `POST /1.0/ubuntu-pro/token` should retrieve a guest token via the Ubuntu Pro client.
+	// - When the guest has this setting, the pro client inside the guest will attempt to retrieve a guest token at startup.
+	GuestAttachSettingOn GuestAttachSetting = "on"
+)
+
+// isValid returns an error if the GuestAttachSetting is not one of the pre-defined values.
+func (g GuestAttachSetting) isValid() error {
+	if !shared.ValueInSlice(g, []GuestAttachSetting{GuestAttachSettingOff, GuestAttachSettingAvailable, GuestAttachSettingOn}) {
+		return fmt.Errorf("Invalid guest auto-attach setting %q", g)
+	}
+
+	return nil
+}
+
+// Settings contains Ubuntu Pro settings relevant to LXD.
+type Settings struct {
+	GuestAttach GuestAttachSetting `json:"guest_attach"`
+}
+
+// ubuntuAdvantageDirectory is the base directory for Ubuntu Pro related configuration.
+const ubuntuAdvantageDirectory = "/var/lib/ubuntu-advantage"
+
+// Client is our wrapper for Ubuntu Pro configuration and the Ubuntu Pro CLI.
+type Client struct {
+	guestAttachSetting GuestAttachSetting
+	monitor            fsmonitor.FSMonitor
+	pro                pro
+}
+
+// pro is an internal interface that is used for mocking calls to the pro CLI.
+type pro interface {
+	getProAPIGuestTokenDataAttributes(ctx context.Context) (*ProAPIGetGuestTokenV1DataAttributes, error)
+}
+
+// proCLI calls the actual Ubuntu Pro CLI to implement the interface.
+type proCLI struct{}
+
+// proAPIGetGuestTokenV1 represents the expected format of calls to `pro api u.pro.attach.guest.get_guest_token.v1`.
+// Not all fields are modelled as they are not required for guest attachment functionality.
+type proAPIGetGuestTokenV1 struct {
+	Result string `json:"result"`
+	Data   struct {
+		Attributes ProAPIGetGuestTokenV1DataAttributes `json:"attributes"`
+	} `json:"data"`
+	Errors []struct {
+		Title string
+	} `json:"errors"`
+}
+
+// ProAPIGetGuestTokenV1DataAttributes contains the expected fields of proAPIGetGuestTokenV1 that must be passed back to
+// the guest for pro attachment to succeed.
+type ProAPIGetGuestTokenV1DataAttributes struct {
```

Comment:
Having "Pro" here seems unnecessary given the whole package is "ubuntupro"?

Reply:
Also "Attributes" seems unnecessary, given that we know its a struct with fields (i.e attributes).

Reply:
Oh I see its because you've named it after the CLI command plus the sub-field you're extracting the data from.